README.md
# therubyrhino

[![Gem Version](https://badge.fury.io/rb/therubyrhino.png)](http://badge.fury.io/rb/therubyrhino)
[![Build Status](https://travis-ci.org/cowboyd/therubyrhino.png?branch=master)](https://travis-ci.org/cowboyd/therubyrhino)
[![Dependency Status](https://gemnasium.com/cowboyd/therubyrhino.png)](https://gemnasium.com/cowboyd/therubyrhino)
[![Join the chat at https://gitter.im/cowboyd/therubyracer](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/cowboyd/therubyracer?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Embed the Mozilla Rhino JavaScript interpreter into Ruby

## REQUIREMENTS:

* JRuby >= 1.6.8

## INSTALL:

`jruby -S gem install therubyrhino`

## FEATURES/PROBLEMS:

* Evaluate JavaScript from with in Ruby
* Embed your Ruby objects into the JavaScript world

## SYNOPSIS:

1. JavaScript goes into Ruby
2. Ruby Objects goes into JavaScript
3. Our shark's in the JavaScript!

```ruby
require 'rhino'
```

* evaluate some simple JavaScript using `eval_js`:
```ruby
eval_js "7 * 6" #=> 42
```

* that's quick and dirty, but if you want more control over your environment,
  use a `Context`:
```ruby
Rhino::Context.open do |context|
  context['foo'] = "bar"
  context.eval('foo') # => "bar"
end
```

* evaluate a Ruby function from JavaScript:
```ruby
Rhino::Context.open do |context|
  context["say"] = lambda {|word, times| word * times}
  context.eval("say("Hello", 3)") #=> HelloHelloHello
end
```

* embed a Ruby object into your JavaScript environment
```ruby
class MyMath
  def plus(a, b)
    a + b
  end
end

Rhino::Context.open do |context|
  context["math"] = MyMath.new
  context.eval("math.plus(20, 22)") #=> 42
end
```

* make a Ruby object be your JavaScript environment
```ruby
math = MyMath.new
Rhino::Context.open(:with => math) do |context|
  context.eval("plus(20, 22)") #=> 42
end

# or the equivalent

math.eval_js("plus(20, 22)")
```

### Context Configuration

* make your standard objects (Object, String, etc...) immutable:
```ruby
Rhino::Context.open(:sealed => true) do |context|
  context.eval("Object.prototype.toString = function() {}") # this is an error!
end
```

* turn on Java integration from JavaScript (probably a bad idea):
```ruby
Rhino::Context.open(:java => true) do |context|
  context.eval("java.lang.System.exit()") # it's dangerous!
end
```

* limit the number of instructions that can be executed to prevent rogue code:
```ruby
Rhino::Context.open(:restrictable => true) do |context|
  context.instruction_limit = 100000
  context.eval("while (true);") # => Rhino::RunawayScriptError
end
```

* limit the time a script executes (rogue scripts):
```ruby
Rhino::Context.open(:restrictable => true, :java => true) do |context|
  context.timeout_limit = 1.5 # seconds
  context.eval %Q{
    for (var i = 0; i < 100; i++) {
      java.lang.Thread.sleep(100);
    }
  } # => Rhino::ScriptTimeoutError
end
```

### Loading JavaScript Source

In addition to just evaluating strings, you can also use streams such as files:

* evaluate bytes read from any File/IO object:
```ruby
  File.open("mysource.js") do |file|
    eval_js file, "mysource.js"
  end
```

* or load it by filename:
```ruby
  Rhino::Context.open do |context|
    context.load("mysource.js")
  end
```

### Configurable Ruby access

By default accessing Ruby objects from JavaScript is compatible with *therubyracer*:
https://github.com/cowboyd/therubyracer/wiki/Accessing-Ruby-Objects-From-JavaScript

Thus you end-up calling arbitrary no-arg methods as if they were JavaScript properties,
since instance accessors (properties) and methods (functions) are indistinguishable:
```ruby
Rhino::Context.open do |context|
  context['Time'] = Time
  context.eval('Time.now')
end
```

However, you can customize this behavior and there's another access implementation
that attempts to mirror only attributes as properties as close as possible:
```ruby
class Foo
  attr_accessor :bar

  def initialize
    @bar = "bar"
  end

  def check_bar
    bar == "bar"
  end
end

Rhino::Ruby::Scriptable.access = :attribute
Rhino::Context.open do |context|
  context['Foo'] = Foo
  context.eval('var foo = new Foo()')
  context.eval('foo.bar') # get property using reader
  context.eval('foo.bar = null') # set property using writer
  context.eval('foo.check_bar()') # called like a function
end
```

If you happen to come up with your own access strategy, just set it directly :
```ruby
Rhino::Ruby::Scriptable.access = FooApp::BarAccess.instance
```

## Safe by default

The Ruby Rhino is designed to let you evaluate JavaScript as safely as possible
unless you tell it to do something more dangerous. The default context is a
hermetically sealed JavaScript environment with only the standard objects and
functions. Nothing from the Ruby world is accessible.

For Ruby objects that you explicitly embed into JavaScript, only the +public+
methods "defined in their classes" are exposed by default e.g.
```ruby
class A
  def a; 'a'; end
end

class B < A
  def b; 'b'; end
end

Rhino::Context.open do |context|
  context['a'] = A.new
  context['b'] = B.new
  context.eval("a.a()") # => 'a'
  context.eval("b.b()") # => 'b'
  context.eval("b.a()") # => 'TypeError: undefined property 'a' is not a function'
end
```

## Context Customizations

Just like the JVM packaged Rhino scripting engine, therubyrhino gem supports
specifying JavaScript context properies (optimization level and language version)
using system properties e.g. to force interpreted mode :

  `jruby -J-Drhino.opt.level=-1 -rtherubyrhino -S ...`

You might also set these programatically as a default for all created contexts :
```ruby
Rhino::Context.default_optimization_level = 1
Rhino::Context.default_javascript_version = 1.6
```

Or using plain old JAVA_OPTS e.g. when setting JavaScript version :

  `-Drhino.js.version=1.7`

## Rhino

Rhino is currently maintained at https://github.com/mozilla/rhino
Release downloads are available at http://www.mozilla.org/rhino/download.html
Rhino is licensed under the MPL 1.1/GPL 2.0 license.

### Using a custom Rhino version

Officially supported versions of Rhino's _js.jar_ are packaged separately as
**therubyrhino_jar** gem. Make sure you're using the latest gem version if you
feel like missing something available with Rhino. For experimenters the jar can
be overriden by defining a `Rhino::JAR_PATH` before `require 'rhino'` e.g. :
```ruby
module Rhino
  JAR_PATH = File.expand_path('lib/rhino/build/rhino1_7R5pre/js.jar')
end
# ...
require 'rhino'
```

## LICENSE:

(The MIT License)

Copyright (c) 2009-2015 Charles Lowell

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md
## jar-1.7.7 2015-09-16

* updated to Mozilla's Rhino 1.7R7 release from 2015-06-17, notes:
  https://github.com/mozilla/rhino/releases/tag/Rhino1_7_7_RELEASE

## jar-1.7.6 2015-04-29

* updated to Mozilla's Rhino 1.7R5 release from 2015-04-15, notes:
  https://github.com/mozilla/rhino/releases/tag/Rhino1_7_6_RELEASE

## jar-1.7.5 2015-04-29

* updated to Mozilla's Rhino 1.7R5 release from 2015-01-29, notes:
  https://github.com/mozilla/rhino/releases/tag/Rhino1_7R5_RELEASE

## 2.0.3 2014-02-20

* minor fix for time_to_js when scope not set (#27)
* Jar path now avoids using a glob, which doesn't work inside a jar (#26)
* use Rhino.warn explicitly otherwise it's a Kernel.warn

## 2.0.2 2012-12-05

* handle Rhino's 64K code generation (method) limit on the fly (#23)
* correct explicit Ruby equality == and eql? (JRuby 1.7.1 compat)
* allow to set javascript version programatically - globally also allow
  reading it from system properties by default
* allow to set optimization level globally also allow reading it from
  system properties by default
* make sure Ruby function wrapper has (Ruby) #call semantics just like
  JavaScript functions exposed into the Ruby side
* function's return value should be converted to Ruby (Ruby #call style)

## 2.0.1 2012-08-24

* JSError improvement to preserve nested Ruby error message
* jar-1.7.4 regression fix e.g when loading less (#25)
* error.message should be a String value (1.9.3 compat)

## jar-1.7.4 2012-08-02

* updated to new Mozilla Rhino 1.7R4 release, notes:
  https://developer.mozilla.org/en/New_in_Rhino_1.7R4

## 2.0.0 2012-08-02

This release is functionally the same as therubyrhino-1.73.5 as long
as the therubyrhino_jar-1.7.3 gem dependency is used along with it.

* new versioning scheme - old scheme used for the jar gem
* moved out the rhino.jar into a separate therubyrhino_jar gem

## jar-1.7.3 2012-08-02

* therubyrhino_jar gem packaged with Mozilla Rhino 1.7R3


## 1.73.5 2012-07-25

* #to_s functionName typo fix for org.mozilla.javascript.ScriptStackElement
* make sure thrown values are correctly raised from inside JS functions
* a better #inspect for native rhino objects
* correct JavaScript error handling for Function#apply

## 1.73.4 2012-05-21

* allow rhino.jar path overrides with Rhino::JAR_PATH
* 'correct' JSError#inspect - show thrown value
* fix JSError#javascript_backtrace to be an array and add it on top of the
  (ruby) backtrace
* make sure JSError#cause always points to native rhino cause (#19)
* avoid using instance variables with 'native' JS::Context (JRuby 1.7 warnings)

## 1.73.3 2012-04-23

RedJS 0.4 compatible

* allow try-catch-ing ScriptError (besides StandardError) in JS
* support for yield in JS property access via the [], []= methods
* refactor access implementations to classes + introduce a shared base
* missing explicit require 'rhino/version'
* Ruby StandardError wrapping so they can be try-catched as JS "error" values
* Rhino::Context.new is expected to yield when block passed

## 1.73.2 2012-04-11

RedJS 0.2.1 compatible

* improve JSError#message + add JSError#value to reflect throw JS value
* correctly convert hashes nested within arrays to_javascript (#12)
* full jruby --1.9 compatibility
* Context.default_factory - no longer use a new factory per context
* restrictable limits now require Contex.new(:restrictable => true)
* added Context#timeout_limit (to complete instruction_limit)

## 1.73.1 2011-11-28

NOTE: this is a "major" code update from 1.73.0 with some incompatibilities
although keeping the bits backward compatible as much as possible :

* add a JS:Function#apply to be used for calling functions from Ruby
* add a JS:Function#bind usable with JS functions from Ruby
* JS:Function#call should work similar to Method/Proc#call
* customizable scriptable access module for resolving Ruby properties from JS
  with TRR compatible Ruby::DefaultAccess and a custom Ruby::AttributeAccess
* implement JavaScript function style argument filling/slicing (for Ruby)
* delegate to hash like method []/[]= when wrapped subject supports them
* introduce Ruby::Constructor JS wrapper for Ruby classes
* hande int JavaScript property resolution
* make sure Time -> Date conversion happens as well
* deprecate JavascriptError use JSError instead with a javascript_backtrace fix
* avoid using Rhino::To.javascript instead use Rhino.to_javascript etc.
* NativeObject/NativeFunction got removed to avoid wrapping - instead Rhino's
  "native" Java classes are customized using JRuby's Java integration
* support for setting JS version via Context.version
* Rhino::J gets deprecated it's now know as Rhino::JS

## 1.73.0 2011-11-28

* upgrade to rhino-1.7R3
* cache objects passed to context - same object passed twice ends up the same
* properly map ruby Time objects to javascript Date
* jruby --1.9 improvements

## 1.72.8 2011-06-26

* fix passing of options hash to ruby.

## 1.72.6 2009-11-30

* 2 major enhancements:
  * evaluate an IO object such as a file or an socket input stream
  * load javascript sources directly into the context with the file system

## 1.72.5 2009-11-12

* 2 major enhancements:
  * evaluate javascript with a ruby object as it's scope using Context#open(:with => object)
  * add eval_js() method to Object to evaluate in the context of that object

## 1.72.4 2009-11-12

* 3 major enhancements:
  * automatically wrap/unwrap ruby and javascript arrays
  * automatically convert ruby method objects and Proc objects into javascript functions
  * Make functions defined in javascript callable from ruby

## 1.72.3 2009-11-11

* 4 major enhancements:
  * greatly simplified interface to context by unifying context and scope
  * remove Context#open_std()
  * remove Context#standard
  * remove Context#init_standard_objects

## 1.72.2 2009-11-10

* 1 major enhancement:
  * ability to limit the instruction count for a context

## 1.72.1 2009-11-09

* 4 major enhancements:

lib/rhino/context.rb
require 'stringio'

module Rhino

  # ==Overview
  #  All Javascript must be executed in a context which represents the execution environment in
  #  which scripts will run. The environment consists of the standard javascript objects
  #  and functions like Object, String, Array, etc... as well as any objects or functions which
  #  have been defined in it. e.g.
  #
  #   Context.open do |cxt|
  #     cxt['num'] = 5
  #     cxt.eval('num + 5') #=> 10
  #   end
  #
  # == Multiple Contexts.
  # The same object may appear in any number of contexts, but only one context may be executing javascript code
  # in any given thread. If a new context is opened in a thread in which a context is already opened, the second
  # context will "mask" the old context e.g.
  #
  #   six = 6
  #   Context.open do |cxt|
  #     cxt['num'] = 5
  #     cxt.eval('num') # => 5
  #     Context.open do |cxt|
  #       cxt['num'] = 10
  #       cxt.eval('num') # => 10
  #       cxt.eval('++num') # => 11
  #     end
  #     cxt.eval('num') # => 5
  #   end
  #
  # == Notes
  # While there are many similarities between Rhino::Context and Java::org.mozilla.javascript.Context, they are not
  # the same thing and should not be confused.
  class Context

    class << self

      # initalize a new context with a fresh set of standard objects. All operations on the context
      # should be performed in the block that is passed.
      def open(options = {}, &block)
        new(options).open(&block)
      end

      def eval(javascript)
        new.eval(javascript)
      end
      
    end
    
    @@default_factory = nil
    def self.default_factory
      @@default_factory ||= ContextFactory.new
    end
    
    def self.default_factory=(factory)
      @@default_factory = factory
    end
    
    @@default_optimization_level = java.lang.Integer.getInteger('rhino.opt.level')
    def self.default_optimization_level
      @@default_optimization_level
    end
    
    def self.default_optimization_level=(level)
      @@default_optimization_level = level
    end

    @@default_javascript_version = java.lang.System.getProperty('rhino.js.version')
    def self.default_javascript_version
      @@default_javascript_version
    end
    
    def self.default_javascript_version=(version)
      @@default_javascript_version = version
    end

    attr_reader :scope
    
    # Create a new javascript environment for executing javascript and ruby code.
    # * <tt>:sealed</tt> - if this is true, then the standard objects such as Object, Function, Array will not be able to be modified
    # * <tt>:with</tt> - use this ruby object as the root scope for all javascript that is evaluated
    # * <tt>:java</tt> - if true, java packages will be accessible from within javascript
    def initialize(options = {}) #:nodoc:
      factory = options[:factory] || 
        (options[:restrictable] ? RestrictableContextFactory.instance : self.class.default_factory)
      factory.call do |context|
        @native = context
        @global = @native.initStandardObjects(nil, options[:sealed] == true)
        if with = options[:with]
          @scope = Rhino.to_javascript(with)
          @scope.setParentScope(@global)
        else
          @scope = @global
        end
        unless options[:java]
          for package in ["Packages", "java", "javax", "org", "com", "edu", "net"]
            @global.delete(package)
          end
        end
      end
      if optimization_level = options[:optimization_level] || self.class.default_optimization_level
        self.optimization_level = optimization_level
      end
      if javascript_version = options[:javascript_version] || self.class.default_javascript_version
        self.javascript_version = javascript_version
      end
      yield(self) if block_given?
    end
    
    # Returns the ContextFactory used while creating this context.
    def factory
      @native.getFactory
    end
    
    # Read a value from the global scope of this context
    def [](key)
      @scope[key]
    end

    # Set a value in the global scope of this context. This value will be visible to all the
    # javascript that is executed in this context.
    def []=(key, val)
      @scope[key] = val
    end

    # Evaluate a string of javascript in this context:
    # * <tt>source</tt> - the javascript source code to evaluate. This can be either a string or an IO object.
    # * <tt>source_name</tt> - associated name for this source code. Mainly useful for backtraces.
    # * <tt>line_number</tt> - associate this number with the first line of executing source. Mainly useful for backtraces
    def eval(source, source_name = "<eval>", line_number = 1)
      open do
        if IO === source || StringIO === source
          result = @native.evaluateReader(@scope, IOReader.new(source), source_name, line_number, nil)
        else
          result = @native.evaluateString(@scope, source.to_s, source_name, line_number, nil)
        end
        Rhino.to_ruby(result)
      end
    end
    
    def evaluate(*args) # :nodoc:
      eval(*args) # an alias
    end

    # Read the contents of <tt>filename</tt> and evaluate it as javascript. Returns the result of evaluating the
    # javascript. e.g.
    #
    # Context.open do |cxt|
    #   cxt.load("path/to/some/lib.js")
    # end
    #
    def load(filename)
      File.open(filename) do |file|
        evaluate file, filename, 1
      end
    end

    # Returns true if this context supports restrictions.
    def restrictable?
      @native.is_a?(RestrictableContextFactory::Context)
    end
    
    def instruction_limit
      restrictable? ? @native.instruction_limit : false
    end
    
    # Set the maximum number of instructions that this context will execute.
    # If this instruction limit is exceeded, then a #Rhino::RunawayScriptError
    # will be raised.
    def instruction_limit=(limit)
      if restrictable?
        @native.instruction_limit = limit
      else
        raise "setting an instruction_limit has no effect on this context, use " + 
              "Context.open(:restrictable => true) to gain a restrictable instance"
      end
    end

    def timeout_limit
      restrictable? ? @native.timeout_limit : false
    end
    
    # Set the duration (in seconds e.g. 1.5) this context is allowed to execute.
    # After the timeout passes (no matter if any JS has been evaluated) and this
    # context is still attempted to run code, a #Rhino::ScriptTimeoutError will
    # be raised.
    def timeout_limit=(limit)
      if restrictable?
        @native.timeout_limit = limit
      else
        raise "setting an timeout_limit has no effect on this context, use " + 
              "Context.open(:restrictable => true) to gain a restrictable instance"
      end
    end
    
    def optimization_level
      @native.getOptimizationLevel
    end
    
    # Set the optimization level that this context will use. This is sometimes necessary
    # in Rhino, if the bytecode size of the compiled javascript exceeds the 64KB limit.
    # By using the -1 optimization level, you tell Rhino to run in interpretative mode,
    # taking a hit to performance but escaping the Java bytecode limit.
    def optimization_level=(level)
      if JS::Context.isValidOptimizationLevel(level)
        @native.setOptimizationLevel(level)
        level
      else
        @native.setOptimizationLevel(0)
        nil
      end
    end

    # Get the JS interpreter version. 
    # Returns a number e.g. 1.7, nil if unknown and 0 for default.
    def javascript_version
      case const_value = @native.getLanguageVersion
        when -1 then nil # VERSION_UNKNOWN
        when  0 then 0 # VERSION_DEFAULT
        else const_value / 100.0 # VERSION_1_1 (1.1 = 110 / 100)
      end
    end
    alias :version :javascript_version
    
    # Sets interpreter mode a.k.a. JS language version e.g. 1.7 (if supported).
    def javascript_version=(version)
      const = version.to_s.gsub('.', '_').upcase
      const = "VERSION_#{const}" if const[0, 7] != 'VERSION'
      if JS::Context.constants.find { |c| c.to_s == const }
        const_value = JS::Context.const_get(const)
        @native.setLanguageVersion(const_value)
        const_value
      else
        @native.setLanguageVersion(JS::Context::VERSION_DEFAULT)
        nil
      end
    end
    alias :version= :javascript_version= 
    
    # Enter this context for operations. 
    # Some methods such as eval() will fail unless the context is open.
    def open(&block)
      do_open(&block)
    rescue JS::RhinoException => e
      if code_generation_error?(e)
        Rhino.warn "[INFO] Rhino byte-code generation failed forcing #{@native} into interpreted mode"
        self.optimization_level = -1
        retry
      end
      raise Rhino::JSError.new(e)
    end
    
    private
    
    def do_open # :nodoc
      factory.enterContext(@native)
      begin
        yield self
      ensure
        factory.exit
      end
    end
    
    CODE_GENERATION_ERROR_MESSAGE = 'generated bytecode for method exceeds 64K limit' # :nodoc

    CODE_GENERATION_TRACE_CLASS_NAME = 'org.mozilla.javascript.optimizer.Codegen' # :nodoc
    CODE_GENERATION_TRACE_METHOD_NAME = 'reportClassFileFormatException' # :nodoc
    # at org.mozilla.javascript.optimizer.Codegen.reportClassFileFormatException

    def code_generation_error?(exception) # :nodoc
      if ( exception.is_a?(NativeException) rescue nil ) # JRuby 1.6 wrapping
        exception = exception.cause
      end
      if exception.class == Rhino::JS::EvaluatorException
        if exception.message.index(CODE_GENERATION_ERROR_MESSAGE)
          return true
        end
        # NOTE: unfortunately Rhino localizes the error messages!
        # and the ClassFileFormatException is not kept as a cause
        class_name = CODE_GENERATION_TRACE_CLASS_NAME
        method_name = CODE_GENERATION_TRACE_METHOD_NAME
        for trace in exception.getStackTrace()
          if class_name == trace.class_name && method_name == trace.method_name
            return true
          end
        end
      end
      false
    end

  end

  class IOReader < java.io.Reader # :nodoc:

    def initialize(io)
      @io = io
    end

    # implement int Reader#read(char[] buffer, int offset, int length)
    def read(buffer, offset, length)
      str = nil
      begin
        str = @io.read(length)
      rescue StandardError => e
        raise java.io.IOException.new("failed reading from ruby IO object")
      end
      if str.nil?
        return -1
      else
        jstr = str.to_java
        for i in 0 .. jstr.length - 1
          buffer[i + offset] = jstr.charAt(i)
        end
        return jstr.length
      end
    end
    
  end

  ContextFactory = JS::ContextFactory # :nodoc: backward compatibility

  class RestrictableContextFactory < ContextFactory # :nodoc:

    @@instance = nil
    def self.instance
      @@instance ||= new
    end
    
    # protected Context makeContext()
    def makeContext
      Context.new(self)
    end
    
    # protected void observeInstructionCount(Context context, int instructionCount)
    def observeInstructionCount(context, count)
      context.check!(count) if context.is_a?(Context)
    end
    
    # protected Object doTopCall(Callable callable, Context context, 
    #                            Scriptable scope, Scriptable thisObj, Object[] args)
    def doTopCall(callable, context, scope, this, args)
      context.reset! if context.is_a?(Context)
      super
    end
    
    class Context < JS::Context # :nodoc:

      def initialize(factory)
        super(factory)
        reset!
      end

      attr_reader :instruction_limit
      
      def instruction_limit=(limit)
        treshold = getInstructionObserverThreshold
        if limit && (treshold == 0 || treshold > limit)
          setInstructionObserverThreshold(limit)
        end
        @instruction_limit = limit
      end

      attr_reader :instruction_count
      
      TIMEOUT_INSTRUCTION_TRESHOLD = 42
      
      attr_reader :timeout_limit
      
      def timeout_limit=(limit) # in seconds
        treshold = getInstructionObserverThreshold
        if limit && (treshold == 0 || treshold > TIMEOUT_INSTRUCTION_TRESHOLD)
          setInstructionObserverThreshold(TIMEOUT_INSTRUCTION_TRESHOLD)
        end
        @timeout_limit = limit
      end
      
      attr_reader :start_time
      
      def check!(count = nil)
        @instruction_count += count if count
        check_instruction_limit!
        check_timeout_limit!(count)
      end

      def check_instruction_limit!
        if instruction_limit && instruction_count > instruction_limit
          raise RunawayScriptError, "script exceeded allowable instruction count: #{instruction_limit}"
        end
      end
      
      def check_timeout_limit!(count = nil)
        if timeout_limit
          elapsed_time = Time.now.to_f - start_time.to_f
          if elapsed_time > timeout_limit
            raise ScriptTimeoutError, "script exceeded timeout: #{timeout_limit} seconds"
          end
          # adapt instruction treshold as needed :
          if count
            treshold = getInstructionObserverThreshold
            if elapsed_time * 2 < timeout_limit
              next_treshold_guess = treshold * 2
              if instruction_limit && instruction_limit < next_treshold_guess
                setInstructionObserverThreshold(instruction_limit)
              else
                setInstructionObserverThreshold(next_treshold_guess)
              end
            end
          end
        end
      end
      
      def reset!
        @instruction_count = 0
        @start_time = Time.now
        self
      end
      
    end
    
  end
    
  class ContextError < StandardError # :nodoc:
  end

  class RunawayScriptError < ContextError # :nodoc:
  end

  class ScriptTimeoutError < ContextError # :nodoc:
  end
  
end
